chromium/browser_type.go

```diff
@@ -33,41 +33,34 @@ var _ api.BrowserType = &BrowserType{}
 // BrowserType provides methods to launch a Chrome browser instance or connect to an existing one.
 // It's the entry point for interacting with the browser.
 type BrowserType struct {
-	Ctx             context.Context
-	CancelFn        context.CancelFunc
-	hooks           *common.Hooks
-	fieldNameMapper *common.FieldNameMapper
-	vu              k6modules.VU
-
-	execPath string       // path to the Chromium executable
-	storage  *storage.Dir // stores temporary data for the extension and user
+	// FIXME: This is only exported because testBrowser needs it. Contexts
+	// shouldn't be stored on structs if we can avoid it.
+	Ctx       context.Context
+	vu        k6modules.VU
+	hooks     *common.Hooks
+	k6Metrics *k6ext.CustomMetrics
+	execPath  string       // path to the Chromium executable
+	storage   *storage.Dir // stores temporary data for the extension and user
 }
 
-// NewBrowserType returns a new Chrome browser type.
-// Before returning a new browser type:
-// - Initializes the extension-wide context
-// - Initializes the goja runtime.
-func NewBrowserType(ctx context.Context) api.BrowserType {
+// NewBrowserType registers our custom k6 metrics, creates method mappings on
+// the goja runtime, and returns a new Chrome browser type.
+func NewBrowserType(vu k6modules.VU) api.BrowserType {
 	var (
-		vu    = k6ext.GetVU(ctx)
 		rt    = vu.Runtime()
 		hooks = common.NewHooks()
 	)
 
-	// Create the extension master context.
-	// If this context is cancelled we'll initiate an extension wide cancellation and shutdown.
-	extensionCtx, extensionCancelFn := context.WithCancel(ctx)
-	extensionCtx = common.WithHooks(extensionCtx, hooks)
-
+	// NOTE: vu.InitEnv() *must* be called from the script init scope,
+	// otherwise it will return nil.
+	k6m := k6ext.RegisterCustomMetrics(vu.InitEnv().Registry)
```

Comment:
The downside of doing this here is that it would have to be done in every `BrowserType` implementation, when it really should be done once for the whole extension.

I moved it here since we need to re-initialize the context on each iteration, and since we store these metrics on the context, we have to re-attach it every time.

An alternative would be to keep this in `NewModuleInstance` and pass it to this and every constructor that needs it, but the signature would be messier. We'll have to do it eventually though, once we add a second `BrowserType` implementation, so let's rethink it then.